File: python-oracle/src/python_oracle/testcontainers_sample/oracle_database_container.py
import oracledb
from testcontainers.core.container import DockerContainer
from testcontainers.core.wait_strategies import LogMessageWaitStrategy


APP_USER = "testuser"
APP_USER_PASSWORD = "welcome12345"

# OracleDatabaseContainer implements a Testcontainers DbContainer for the
# gvenzl/oracle-free image variants.
class OracleDatabaseContainer(DockerContainer):
    def __init__(self,
                 app_user=APP_USER,
                 app_user_password=APP_USER_PASSWORD,
                 image="gvenzl/oracle-free:23.26.0-slim-faststart",
                 db_name="freepdb1",
                 container_port=1521,
                 host="localhost",
                 **kwargs):
        super(OracleDatabaseContainer, self).__init__(image=image, **kwargs)
        self.container_port = container_port
        self.app_user = app_user
        self.app_user_password = app_user_password
        self.db_name=db_name
        self.host = host

        # Configure the container via environment variables
        self.with_env("ORACLE_RANDOM_PASSWORD", "y")
        self.with_env("APP_USER", app_user)
        self.with_env("APP_USER_PASSWORD", app_user_password)
        # Database listens on 1521 by default
        self.with_exposed_ports(container_port)
        # Wait for the database to be ready
        self.waiting_for(LogMessageWaitStrategy("DATABASE IS READY TO USE!"))

    def get_connection(self) -> oracledb.Connection:
        bind_port = self.get_exposed_port(self.container_port)
        return oracledb.connect(user=self.app_user, password=self.app_user_password,
                                host=self.host, port=bind_port, service_name=self.db_name)

    def get_connection_url(self) -> str:
        bind_port = self.get_exposed_port(self.container_port)
        return f'oracle+oracledb://{self.app_user}:{self.app_user_password}@{self.host}:{bind_port}/?service_name={self.db_name}'

    def _configure(self):
        pass

File: typescript/src/testcontainers/oracle_database_container.ts
import {GenericContainer, type StartedTestContainer, Wait} from "testcontainers";
import * as OracleDB from "oracledb";
import type {Connection} from "oracledb";

// Implements an Oracle Database Free container for Testcontainers.
// Run the sample using "npm run testcontainers-example"
export class OracleDatabaseContainer {
    private container?: StartedTestContainer | undefined;

    constructor(
        private readonly image: string = "gvenzl/oracle-free:23.9-slim-faststart",
        private readonly port: number = 1521,
        private readonly username: string = "testuser",
        private readonly password: string = "Welcome12345",
        private readonly serviceName: string = "freepdb1"
    ) {}

    public async start(): Promise<StartedTestContainer> {
        this.container = await new GenericContainer(this.image)
            .withExposedPorts(this.port)
            .withEnvironment({
                "ORACLE_RANDOM_PASSWORD": "y",
                "APP_USER": this.username,
                "APP_USER_PASSWORD": this.password
            })
            .withWaitStrategy(Wait.forLogMessage("DATABASE IS READY TO USE!"))
            .start();

        return this.container;
    }

    public getHost(): string {
        if (!this.container) throw new Error("Container not started yet");
        return this.container.getHost();
    }

    public getPort(): number {
        if (!this.container) throw new Error("Container not started yet");
        return this.container.getMappedPort(this.port);
    }

    public async getDatabaseConnection(): Promise<Connection> {
        return OracleDB.getConnection({
            user: this.username,
            password: this.password,
            connectionString: this.getConnectionString()
        })
    }

    public getConnectionString(): string {
        return `${this.getHost()}:${this.getPort()}/${this.serviceName}`;
    }

    public async stop(): Promise<void> {
        if (this.container) {
            await this.container.stop();
            this.container = undefined;
        }
    }
}
